src/iosrv/sched.c

```diff
@@ -0,0 +1,542 @@
+		D_ASSERT(cycle->sc_ults_tot == 0);
+		return false;
+	}
+
+	/* Need nvme poll to end current cycle */
+	if (cycle->sc_ults_tot == 0)
+		return true;
+
+	/*
+	 * Need extra NVMe poll when too many ULTs are processed in
+	 * current cycle.
+	 */
+	if (cycle->sc_age_nvme > cycle->sc_age_nvme_bound[1])
+		return true;
+
+	/* TODO: Take NVMe I/O statistics into account */
+
+	return false;
+}
+
+static ABT_unit
+sched_pop_nvme_poll(struct sched_data *data, ABT_pool pool)
+{
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+	ABT_unit		 unit;
+	int			 ret;
+
+	if (!need_nvme_poll(cycle))
+		return ABT_UNIT_NULL;
+
+	D_ASSERT(cycle->sc_cycle_started);
+	cycle->sc_age_nvme = 0;
+	cycle->sc_age_net++;
+	if (cycle->sc_ults_tot == 0)
+		cycle->sc_cycle_started = 0;
+
+	/* Only main xstream (VOS xstream) has NVMe poll ULT */
+	if (!dx->dx_main_xs)
+		return ABT_UNIT_NULL;
+
+	ret = ABT_pool_pop(pool, &unit);
+	if (ret != ABT_SUCCESS) {
+		D_ERROR("XS(%d) failed to pop NVMe poll ULT: %d\n",
+			dx->dx_xs_id, ret);
+		return ABT_UNIT_NULL;
+	}
+
+	return unit;
+}
+
+static ABT_unit
+sched_pop_one(struct sched_data *data, ABT_pool pool, int pool_idx)
+{
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+	ABT_unit		 unit;
+	int			 ret;
+
+	D_ASSERT(cycle->sc_ults_tot >= cycle->sc_ults_cnt[pool_idx]);
+	if (cycle->sc_ults_cnt[pool_idx] == 0)
+		return ABT_UNIT_NULL;
+
+	ret = ABT_pool_pop(pool, &unit);
+	if (ret != ABT_SUCCESS) {
+		D_ERROR("XS(%d) failed to pop ULT for ABT pool(%d): %d\n",
+			dx->dx_xs_id, pool_idx, ret);
+		return ABT_UNIT_NULL;
+	}
+
+	/* XXX Need to figure out why it can pop a NULL unit */
+	if (unit == ABT_UNIT_NULL)
+		D_ERROR("XS(%d) poped NULL unit for ABT pool(%d)\n",
+			dx->dx_xs_id, pool_idx);
+
+	cycle->sc_age_net++;
+	cycle->sc_age_nvme++;
+	cycle->sc_ults_cnt[pool_idx] -= 1;
+	cycle->sc_ults_tot -= 1;
+
+	return unit;
+}
+
+static void
+sched_start_cycle(struct sched_data *data, ABT_pool *pools)
+{
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+	size_t			 cnt;
+	uint32_t		 limit = 0;
+	bool			 space_pressure = false;
+	int			 i, ret;
+
+	D_ASSERT(cycle->sc_new_cycle == 1);
+	D_ASSERT(cycle->sc_cycle_started == 0);
+	D_ASSERT(cycle->sc_ults_tot == 0);
+
+	cycle->sc_new_cycle = 0;
+	cycle->sc_cycle_started = 1;
+
+	/* Get number of ULTS for each ABT pool */
+	for (i = DSS_POOL_IO; i < DSS_POOL_CNT; i++) {
+		D_ASSERT(cycle->sc_ults_cnt[i] == 0);
+
+		ret = ABT_pool_get_size(pools[i], &cnt);
+		if (ret != ABT_SUCCESS) {
+			D_ERROR("XS(%d) get ABT pool(%d) size error: %d\n",
+				dx->dx_xs_id, i, ret);
+			cnt = 0;
+		}
+
+		cycle->sc_ults_cnt[i] = cnt;
+		cycle->sc_ults_tot += cycle->sc_ults_cnt[i];
+
+		if (sched_throttle[i] > 0 && cycle->sc_ults_cnt[i] > 1)
+			limit = 1;
+	}
+
+	/* No throttling for helper xstream so far */
+	if (!dx->dx_main_xs)
+		return;
+
+	if (!limit && !space_pressure)
```

Comment:
probably if the system is idle, i.e. no I/O inflight(rpc ULT), it should not throttle as well?

Reply:
Yes, check the code below.